### PR TITLE
feat: remove illegal character check

### PR DESCRIPTION
**Pull request by chalmerlowe**
This PR provides the ability to incorporate additional characters in labels and similar identifiers. 

There are a number of characters allowed in BigQuery identifiers that get filtered out in the existing code. The PR prevents that from happening.

A list of the special characters is included [here](https://cloud.google.com/bigquery/docs/schemas#flexible-column-names):

```
An ampersand (&)
A percent sign (%)
An equals sign (=)
A plus sign (+)
A colon (:)
An apostrophe (')
A less-than sign (<)
A greater-than sign (>)
A number sign (#)
A vertical line (|)
```

Fixes an internal buganizer issue: [🦕](https://buganizer.corp.google.com/issues/291076163)

### CODE REVIEW

File: sqlalchemy_bigquery/base.py [hunk 1]
```diff
@@ -106,8 +112,9 @@ def format_label(self, label, name=None):
             name = "_" + name
 
         # Fields must contain only letters, numbers, and underscores
-        name = FIELD_ILLEGAL_CHARACTERS.sub("_", name)
+        # name = FIELD_ILLEGAL_CHARACTERS.sub("_", name)
```
Comment (chalmerlowe) [inline sqlalchemy_bigquery/base.py:115]:
```suggestion
        name = FIELD_ILLEGAL_CHARACTERS.sub("_", name)
```

File: sqlalchemy_bigquery/base.py [hunk 2]
```diff
@@ -56,7 +56,7 @@
 from .parse_url import parse_url
 from . import _helpers, _struct, _types
 
-FIELD_ILLEGAL_CHARACTERS = re.compile(r"[^\w]+")
+FIELD_ILLEGAL_CHARACTERS = re.compile(r"[^\w]+", re.ASCII)
```
Comment (chalmerlowe) [inline sqlalchemy_bigquery/base.py:59]:
```suggestion
FIELD_ILLEGAL_CHARACTERS = re.compile(r"[^\w.]+", re.ASCII)
```

File: sqlalchemy_bigquery/base.py [hunk 3]
```diff
@@ -56,7 +56,7 @@
 from .parse_url import parse_url
 from . import _helpers, _struct, _types
 
-FIELD_ILLEGAL_CHARACTERS = re.compile(r"[^\w]+")
+FIELD_ILLEGAL_CHARACTERS = re.compile(r"[^\w.]+", re.ASCII)
```
Comment (chalmerlowe) [inline sqlalchemy_bigquery/base.py:59]:
```suggestion
FIELD_ILLEGAL_CHARACTERS = re.compile(r"[^\w&%=+:'<>#|]+", re.ASCII)
```

File: sqlalchemy_bigquery/base.py [hunk 4]
```diff
@@ -98,6 +98,12 @@ def quote(self, ident, force=None, column=False):
         else:
             return ident
 
+    def backtick_quote(self, name):
+        """
+        Encapsulates statements with backticks.
+        """
+        return f"`{name}`"
+
```
Comment (chalmerlowe) [inline sqlalchemy_bigquery/base.py:106]:
```suggestion
```

File: sqlalchemy_bigquery/base.py [hunk 5]
```diff
@@ -56,7 +56,9 @@
 from .parse_url import parse_url
 from . import _helpers, _struct, _types
 
-FIELD_ILLEGAL_CHARACTERS = re.compile(r"[^\w]+")
+# Incorporates special characters used in flexible column names:
+# https://cloud.google.com/bigquery/docs/schemas#flexible-column-names
+FIELD_ILLEGAL_CHARACTERS = re.compile(r"[^\w&%=+:'<>#|]+", re.ASCII)
```
Comment (tswast) [inline sqlalchemy_bigquery/base.py:61]: Whitespace is allowed now, right? I used `[^!"$()*,./;?@[\\\]^`{}~\n]*` in https://github.com/ibis-project/ibis/pull/6557 and it seems to be working alright.

Comment (chalmerlowe) [inline sqlalchemy_bigquery/base.py:61]:
```suggestion
FIELD_ILLEGAL_CHARACTERS = re.compile(r"[^\w&%=+:'<>#|\n]+", re.ASCII)
```

File: sqlalchemy_bigquery/base.py [hunk 6]
```diff
@@ -56,7 +56,9 @@
 from .parse_url import parse_url
 from . import _helpers, _struct, _types
 
-FIELD_ILLEGAL_CHARACTERS = re.compile(r"[^\w]+")
+# Incorporates special characters used in flexible column names:
```
Comment (chalmerlowe) [inline sqlalchemy_bigquery/base.py:59]:
```suggestion
# Illegal characters is intended to be all characters that are not explicitly
# allowed as part of the flexible column names. 
```

File: sqlalchemy_bigquery/base.py [hunk 7]
```diff
@@ -56,7 +56,10 @@
 from .parse_url import parse_url
 from . import _helpers, _struct, _types
 
-FIELD_ILLEGAL_CHARACTERS = re.compile(r"[^\w]+")
+# Illegal characters is intended to be all characters that are not explicitly
+# allowed as part of the flexible column names.
+# https://cloud.google.com/bigquery/docs/schemas#flexible-column-names
+FIELD_ILLEGAL_CHARACTERS = re.compile(r"[^\w&%=+:'<>#|\n]+", re.ASCII)
```
Comment (tswast) [inline sqlalchemy_bigquery/base.py:62]: Let's just list out the characters that aren't allowed in https://cloud.google.com/bigquery/docs/schemas#flexible-column-names

```suggestion
FIELD_ILLEGAL_CHARACTERS = re.compile(r"[!"$()*,./;?@[\\\]^{}~\n]+", re.ASCII)
```

File: sqlalchemy_bigquery/base.py [hunk 8]
```diff
@@ -56,7 +56,10 @@
 from .parse_url import parse_url
 from . import _helpers, _struct, _types
 
-FIELD_ILLEGAL_CHARACTERS = re.compile(r"[^\w]+")
+# Illegal characters is intended to be all characters that are not explicitly
+# allowed as part of the flexible column names.
+# https://cloud.google.com/bigquery/docs/schemas#flexible-column-names
+FIELD_ILLEGAL_CHARACTERS = re.compile(r"[!"$()*,./;?@[\\\]^{}~\n]+", re.ASCII)
```
Comment (chalmerlowe) [inline sqlalchemy_bigquery/base.py:62]:
```suggestion
FIELD_ILLEGAL_CHARACTERS = re.compile(r'[!"$()*,./;?@[\\\]^{}~\n]+', re.ASCII)
```

**Review by tswast — APPROVED** [review]

**Pull request merged**